### ISSUE TITLE
JavaParser not handling a failed solve correctly when using CombinedTypeSolver

### ISSUE THREAD

**Issue by HoldYourWaffle** [javaparser/javaparser#1946]
Please forgive me for the jumbled title. I have no idea how to describe this in a single sentence.
Related to #1942 and (I think) #1945.

I have run into another problem while parsing my source code. Starting to wonder if it's my code being bad, JSS not being stable or me using this completely wrong 😅 

It all started with a method call like this:
```java
UtilityClass.method(SomeClass.InnerEnum.CONSTANT);
```
Which throws the following exception:
```
java.lang.RuntimeException: Unable to calculate the type of a parameter of a method call. Method call: UtilityClass.method(UtilityClass.InnerEnum.CONSTANT), Parameter: UtilityClass.InnerEnum.CONSTANT
    at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:225)
    at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:240)
    at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:125)
    at com.github.javaparser.symbolsolver.JavaSymbolSolver.resolveDeclaration(JavaSymbolSolver.java:142)
    at com.github.javaparser.ast.expr.MethodCallExpr.resolve(MethodCallExpr.java:313)
    at TestJavaParser.testReflectionFirst(TestJavaParser.java:36)
    ... snip ...
Caused by: java.lang.UnsupportedOperationException
    at com.github.javaparser.symbolsolver.reflectionmodel.ReflectionEnumConstantDeclaration.getType(ReflectionEnumConstantDeclaration.java:29)
    at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.solveDotExpressionType(TypeExtractor.java:162)
    at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:181)
    at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
    at com.github.javaparser.ast.expr.FieldAccessExpr.accept(FieldAccessExpr.java:90)
    at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:447)
    at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:310)
    at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveArguments(JavaParserFacade.java:221)
    ... 28 more
```

Thinking back to #1945, I tried parsing the line split-out like this:
```java
SomeClass.InnerEnum parameter = SomeClass.InnerEnum.CONSTANT;
UtilityClass.method(parameter);
```
This resolves perfectly fine. I was creating another sample project with different variants of it to post with this issue, but I disovered some very strange behavior.

I tried using a `JavaParserTypeSolver` in addition to the `ReflectionTypeSolver` (as I learned in #1942, although as far as I know that was caused by dependencies), but that didn't work. I initially assumed that it was a bug related to having an enum-type inside a class-type, but it doesn't work when using a separate enum class either. However, it does seem to always work if I split-out the enum parameter to a temporary variable before passing it to the method.

But then I realized something: the actual cause of the exception is the same `UnsupportedOperationException` as with #1942, which implies that my `ReflectionTypeSolver` was being used instead of the `JavaParserTypeSolver`. This seemed strange to me, because what's the point of having a `CombinedTypeSolver` if it just errors when a try is unsuccesful?
I looked through the source code, and [this](https://github.com/javaparser/javaparser/blob/d5698c6b118d6741dd8d984c37091196b7a0a810/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/typesolvers/CombinedTypeSolver.java#L61) line intrigued me. It looks to me (with very limited JavaParser knowledge) as if it just loops though all the registered solvers, only returning if the solution was succesful. This works in most cases, but `ReflectionTypeSolver` actually *[throws an exception](https://github.com/javaparser/javaparser/blob/d5698c6b118d6741dd8d984c37091196b7a0a810/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/reflectionmodel/ReflectionEnumConstantDeclaration.java#L29)* when it fails to resolve the enum. As far as I can tell this isn't handled properly, which causes the the resolution to completely fail, instead of at least trying to solve it with a secondary resolver.
To test my hypothesis I simply changed my `CombinedTypeSolver` constructor from this:
```java
CombinedTypeSolver typeSolver = new CombinedTypeSolver(new TypeSolver[] {
    new ReflectionTypeSolver(false),
    new JavaParserTypeSolver("src/main/java")
});
```
To this:
```java
CombinedTypeSolver typeSolver = new CombinedTypeSolver(new TypeSolver[] {
    new JavaParserTypeSolver("src/main/java"),
    new ReflectionTypeSolver(false)
});
```
And suddenly every variation resolved just fine.

I have created another [sample project](https://github.com/HoldYourWaffle/JavaParserEnumChoke) which represents my code in a very simplified way.
It consists of a [JUnit test](https://github.com/HoldYourWaffle/JavaParserEnumChoke/blob/master/src/test/java/TestJavaParser.java) that tries to parse [`MainClass.java`](https://github.com/HoldYourWaffle/JavaParserEnumChoke/blob/master/src/main/java/main/MainClass.java). It uses a parameterized unit test to run the test two times: first with the `ReflectionTypeSolver` registered *after* the `JavaParserTypeSolver` (this test should pass), and a second time the other way around (this test currently fails).

I hope my sample project gives enough information to fix this bug. I'd create a PR and do it myself, but I feel like I don't know enough about how JavaParser works to create an elegant solution.
My bonehead attempt using the knowledge I got while investigating would be to do something like this:
```java
// OLD
for (TypeSolver ts : elements) {
    SymbolReference<ResolvedReferenceTypeDeclaration> res = ts.tryToSolveType(name);
    if (res.isSolved()) {
        return res;
    }
}

// NEW
for (TypeSolver ts : elements) {
    try {
        SymbolReference<ResolvedReferenceTypeDeclaration> res = ts.tryToSolveType(name);
        if (res.isSolved()) {
            return res;
        }
    } catch (UnsupportedOperationException e) {
        continue; //try the next resolver
    }
}
```
In case that's how it should be solved (which would surprise me given how complicated this project looks) I'd be happy to create a PR of course.

**Comment by HoldYourWaffle** [comment]
That's a lot better 😉

**Comment by ftomassetti** [comment]
The problem is that JSS is _incomplete_. We report that very clearly by throwing an `UnsupportedOperationException`. In my opinion the most correct way to solve it is implementing the missing functionality.

**Comment by ftomassetti** [comment]
I just submitted a PR for this. In general there are cases where `UnsupportedOperationException` is thrown and they can be straightforward to fix, the only issue is a matter of time. The reflection subsystem is the one that has more of these, as the others are used more frequently.

**Comment by HoldYourWaffle** [comment]
That would of course be the best solution, but that doesn't solve the "overlying" issue (which is caused by the underlying problem, the UOE). Please forgive me for my garbage English :)

The fact that a `CombinedTypeSolver` fails when a single solver doesn't support a part of the solving process kind of defeats it's point in my opinion. What's the point of having multiple (backup) solvers if it doesn't try using them because one of them decided that it doesn't support a part of the process? Catching only a UOE would still surface any actual errors that might occur, but would enable a silent redirect to any (supporting) solvers if there are any.

I also just remembered that the fact that `CombinedTypeSolver` doesn't handle errors correctly has nothing to do with it working perfectly fine if I use a temporary variable. Should I create a separate issue for this?

**Comment by HoldYourWaffle** [comment]
Oh I see that you responded while I was writing my response.
I completely agree that it's a temporary thing, but I still think catching UOE's would solve an underlying (design) issue.

**Comment by ftomassetti** [comment]
Well, the fact is that we really do not want an UnsupportedOperationException to go unnoticed, as it would leave to the false assumption that this thing cannot be solved while it _can_ be, the problem is just with the current implementation. If we swallow the exception we risk to have a bug very difficult to track while here the problem was very obvious.

Regarding the thing with splitting the code in two, I suggest trying this with the new code and see if it is still an issue or not.

**Comment by ftomassetti** [comment]
Also, thank you very much for the detailed bug report. W.r.t. English, I am Italian so your English sounds just perfect to me.

**Comment by HoldYourWaffle** [comment]
> Well, the fact is that we really do not want an UnsupportedOperationException to go unnoticed, as it would leave to the false assumption that this thing cannot be solved while it can be, the problem is just with the current implementation. If we swallow the exception we risk to have a bug very difficult to track while here the problem was very obvious.

That's a really good point. Hadn't thought of that one.

> Regarding the thing with splitting the code in two, I suggest trying this with the new code and see if it is still an issue or not.

I'll try this soon.

> Also, thank you very much for the detailed bug report

Got another one incoming 😉 

> W.r.t. English, I am Italian so your English sounds just perfect to me.

That's a great relief! I feel like most of my English is fairly good, but some sentences are just utterly broken. Great to hear it's not a problem.

**Comment by HoldYourWaffle** [comment]
> Regarding the thing with splitting the code in two, I suggest trying this with the new code and see if it is still an issue or not.

Just tested it, everything passes with the new code.

**Comment by ftomassetti** [comment]
Fixed by #1947 
Thank you for testing it @HoldYourWaffle !
